feature/browzine-sersol-360-adapter-BZ-4334

tiandavis opened the pull request:
## Summary - [BZ-4334](https://thirdiron.atlassian.net/browse/BZ-4334)
Enhances Serial Solutions' 360 Core discovery service's Journal search results with BrowZine. Also, adds the Primo discovery service adapter with a small update to remove the AWS proxy and point directly to the BrowZine Public API.

## Description
Adds a "View Journal in BrowZine" and Journal Cover Image to 360 Core search results when applicable.

Sandbox Search Link (LIVE): http://ae.preview.serialssolutions.com/ejp/?libHash=CS3WE9ZV5H#/search/?searchControl=title&searchType=title_contains&criteria=new%20england%20journal%20of%20medicine&titleType=JOURNALS&filterBy=All&beginPage=0&language=en-US

Credentials: See, JIRA for credentials. https://thirdiron.atlassian.net/browse/BZ-4334
- Visit (Only EVER click the Sandbox link): https://clientcenter.serialssolutions.com/CC/Library/Default.aspx?LibraryCode=M98
- Login
- Click the "Sandbox Search Link" above

![image](https://user-images.githubusercontent.com/376230/35127347-c63474b2-fc7f-11e7-9b81-d5f79f801155.png)

Primo Search Link (LIVE: https://sandbox01-na.alma.exlibrisgroup.com/discovery/search?query=any,contains,new%20england%20journal%20of%20medicine&tab=Everything&search_scope=MyInst_and_CI&vid=01COLSCHL_INST:BRZNTEST&offset=0

## Deploy Precautions
- None

KeeganJ commented on src/browzine-summon-adapter.js:
```diff
@@ -198,34 +198,225 @@ browzine.search = (function() {
+      endpoint += "&access_token=" + apiKey;
+    }
+
+    return endpoint;
```
Just checking this method because we declare `endpoint` without setting it. Are we OK returning `undefined` here if an issn isn't passed in? Some similar methods below (like `getCoverImageUrl`) seem to set the value as null first, so I figured I'd double check the outlier.

tiandavis replied:
Consistency is good. Thanks!

KeeganJ commented on src/browzine-summon-adapter.js:
```diff
@@ -198,34 +198,225 @@ browzine.search = (function() {
+    var scope = getScope(searchResults);
+    var titles = addTargets(getTitles(scope));
+
+    (function poll(titles) {
```
Is there a reason we're wrapping this chunk within the above method in a self executable?

tiandavis replied:
That section needed clarity, Thanks Keegan! Refactored that piece to its own function and added some tests for the new function.

karlbecker commented:
Alas, looks like http://ae.preview.serialssolutions.com/ejp/?libHash=CS3WE9ZV5H#/search/?searchControl=title&searchType=title_contains&criteria=new%20england%20journal%20of%20medicine&titleType=JOURNALS&filterBy=All&beginPage=0&language=en-US is not working here - I get this weird message:
![image](https://user-images.githubusercontent.com/1091559/35065674-868b2596-fb93-11e7-9cd1-f632227d0ae1.png)

I tried just going to like http://ae.preview.serialssolutions.com/ejp/ but that's not working, either - any ideas, @tiandavis ?

KeeganJ commented:
@karlbecker I made that mistake at first too, you have to auth through the instructions further down in the PR first:

> Credentials: See, JIRA for credentials. https://thirdiron.atlassian.net/browse/BZ-4334

> Visit (Only EVER click the Sandbox link): https://clientcenter.serialssolutions.com/CC/Library/Default.aspx?LibraryCode=M98
> Login

After that that page should render correctly.

tiandavis commented:
Yeah, you gotta log into Serial Solutions admin first:
https://clientcenter.serialssolutions.com/CC/Library/Default.aspx?LibraryCode=M98

Unlike Summon, that 360 Core sandbox checks for a session or something.

karlbecker commented:
Interestingly, it works fine after I logged into the admin site - but not before that.  And not in other browsers - only in the browser where I logged into the admin site.

karlbecker commented on tests/acceptance/browzine-sersol-360-core-adapter-test.js:
```diff
@@ -0,0 +1,67 @@
+describe("BrowZine SerSol 360 Core Adapter >", function() {
+  var serSol360Core = {}, searchResults = {};
```
Can we `use strict` so we can use `let` and `const`?

tiandavis replied:
Good question. Actually took that out some time ago so we can keep the test environment as close to the production environment as possible.

karlbecker replied:
No problemo 👍

karlbecker commented on tests/acceptance/browzine-sersol-360-core-adapter-test.js:
```diff
@@ -0,0 +1,67 @@
+          titles: [
+            {
+              title: "The New England journal of medicine",
+              syndeticsImageUrl: "https://secure.syndetics.com/index.aspx?isbn=/mc.gif&issn=0028-4793&client=mistatesum",
```
If this is a URL that actually pings some server that maybe we shouldn't be hitting too frequently from tests, it might be good to just point it at something else, even an assets.thirdiron.com URL for some random cover, even if we label the item as `syndeticsImageUrl`

Perhaps `https://assets.thirdiron.com/default-cover-image.png?this=from-syndetics-i-guess` or something? 😊 

I just wouldn't want Syndetics to contact mistatesum and ask, "Why do we keep getting these random requests from continuous integration servers?" since I don't know what kind of Terms of Service there are for the Syndetics service.

But of course, only if this would have the possibility to truly trigger a request to that URL.  If not, then heyyy no problemo.

tiandavis replied:
hehe, yeah we're definitely not hitting that url, just parsing it. When we do the magic to get the 360 Core results data, we get back an angular scope with a property called `searchResultsCtrl`, so that whole section is just modeling that angular `$scope.searchResultsCtrl` property.

karlbecker replied:
If we don't hit it, then we're fine 👍

karlbecker commented on tests/acceptance/browzine-sersol-360-core-adapter-test.js:
```diff
@@ -0,0 +1,67 @@
+    serSol360Core.adapter(searchResults);
+  });
+
+  describe("search results journal >", function() {
```
🙌  Great to see tests on this!

tiandavis commented on src/browzine-summon-adapter.js:
```diff
@@ -198,34 +198,232 @@ browzine.search = (function() {
   };
 }());
 
+browzine.serSol360Core = (function() {
```
Initial goal was to do it in one file, but thinking it might be best to do separate files because looks like there will be a few Discovery Services to tackle in the future. We could have something like this:

Summon:
https://s3.amazonaws.com/browzine-adapters/summon/browzine-summon-adapter.js

360 Core:
https://s3.amazonaws.com/browzine-adapters/360-core/browzine-360-core-adapter.js

I think that should work for customers. Different products, slightly different link. We'll have to tweak the deployment slightly, but should be doable. We'll need to update the README for each product's loader. e.g. Something like:
```
var browzine = {
  api: "https://api.thirdiron.com/public/v1/libraries/XXX",
  apiKey: "ENTER API KEY",
};

browzine.script = document.createElement("script");
browzine.script.src = "https://s3.amazonaws.com/browzine-adapters/360-core/browzine-360-core-adapter.js";
document.head.appendChild(browzine.script);
```

karlbecker replied:
Yeah - I'm guessing we can't do Primo in this same code base, correct?

If we could bring that over, well wooooo that would be slick 😄

karlbecker replied:
> I think that should work for customers. Different products, slightly different link.

Yeah, I think having slightly different instructions for each one is fine... but I agree, one file to rule them all is pretty sweet 😄

tiandavis replied:
> I'm guessing we can't do Primo in this same code base, correct?

Good question. Not sure TBH, but I'll kick the tires in that next Primo JIRA and see what I find.

karlbecker replied:
Thanks for checking in that next Primo change, Tian!

tiandavis commented on README.md:
```diff
@@ -42,6 +42,7 @@ var browzine = {
   articleWording: "View Complete Issue",
   journalBrowZineWebLinkText: "Browse Now",
```
> but I have a feeling this might have to do with testability of the code
They are different phrases. Summon "Browse Now", 360 Core "View Journal in BrowZine".

`summonJournalBrowZineWebLinkText` kinda want to do this, just concerned if any customers are using the old property, I might be able to layer the use like: `var text = journalBrowZineWebLinkText || summonJournalBrowZineWebLinkText || "default text"` or something to be backwards compatible. Though it's not like a breaking change because we use default text anyway...

Let me think about that.

karlbecker replied:
Yeah, I like the `||` of those strings 👍

tiandavis commented on README.md:
```diff
@@ -1,7 +1,7 @@
 
 ## Overview
-Use `browzine-summon-adapter` to enhance Summon search results with BrowZine data; Adds links to Journal and Article content in BrowZine; Uses BrowZine Journal covers. The BrowZine Summon adapter supports IE11+ and evergreen browsers (Chrome, Firefox, Safari, Microsoft Edge).
+Use `browzine-summon-adapter` to enhance Summon or 360 Core search results with BrowZine data; Adds links to Journal and Article content in BrowZine; Uses BrowZine Journal covers. The BrowZine Summon adapter supports IE11+ and evergreen browsers (Chrome, Firefox, Safari, Microsoft Edge).
```
Two separate core classes: `browzine.summon` and `browzine.serSol360Core` so not so much worried about the core functionality. I do worry about them sharing the same MutationObserver instance however. But, I know we're thinking to have a separate class for each service, so when we do that, that should this particular risk because yes, one could affect the other at the MutationObserver integration point.

karlbecker replied:
Cool - hopefully it's not _too_ much extra work to accomplish!

tiandavis commented:
Validated 360-core and Summon work independently and when deployed together.

https://s3.amazonaws.com/browzine-adapters/summon/browzine-summon-adapter-embed-118.js

```
var browzine = {
  api: "https://api.thirdiron.com/public/v1/libraries/XXX",
  apiKey: "ENTER API KEY",
};

browzine.script = document.createElement("script");
browzine.script.src = "https://s3.amazonaws.com/browzine-adapters/summon/feature/browzine-summon-adapter.js";
document.head.appendChild(browzine.script);

browzine.script = document.createElement("script");
browzine.script.src = "https://s3.amazonaws.com/browzine-adapters/360-core/feature/browzine-360-core-adapter.js";
document.head.appendChild(browzine.script);
```

![image](https://user-images.githubusercontent.com/376230/35158317-b7777f34-fd04-11e7-9513-8e6c1d1392fa.png)

![image](https://user-images.githubusercontent.com/376230/35158359-d3c87e5e-fd04-11e7-8ad3-37f32c4ddfe6.png)

karlbecker commented on README.md:
```diff
@@ -1,16 +1,24 @@
+
+![Summon Journal Search Results](https://i.imgur.com/B34LEec.png "Summon Journal Search Results")
+
+_360 Core Search Results_
```
Nice to see it labeled this way - we've called it SerSol a lot around here, but the labeling on [the official website](http://www.proquest.com/products-services/360-Core.html) is just 360 Core.

tiandavis replied:
Yep! If only, we could name a JavaScript object starting w/ numbers. 😂

karlbecker commented on README.md:
````diff
@@ -22,26 +30,78 @@ browzine.script.src = "https://s3.amazonaws.com/browzine-adapters/summon/browzin
+document.head.appendChild(browzine.script);
+```
+
+_Loading Discovery Service Adapters_
````
I'm not quite following how this is different than the above two adapter scripts... who is the intended audience for this code snippet?  Seems like the 360 Core and Summon code snippets above cover both of these cases.

karlbecker commented on README.md:
```diff
@@ -22,26 +30,78 @@ browzine.script.src = "https://s3.amazonaws.com/browzine-adapters/summon/browzin
 ![Summon 2.0 External Script](https://i.imgur.com/piLMSic.png "Summon 2.0 External Script")
 
+
+_Adding 360 Core Custom Script_
```
It might be a little easier for people new to this repo (like myself!) to follow if we group the instructions together by product.  So organization like this:

* Summon Adapter Installation
  * BrowZine Summon Adapter Script
  * Adding Summon Custom Script
  * Customizations
* 360 Core Adapter Installation
  * BrowZine 360 Core Adapter Script
  * Adding 360 Core Custom Script
  * Customizations

If you'd like me to reorganize it, let me know!

tiandavis replied:
OMG, YES! HEHE. See, the "Loading Multiple Discovery Service Adapters" note above too:
https://github.com/thirdiron/browzine-summon-adapter/pull/15#discussion_r162673571

Thanks @karlbecker!

karlbecker commented on README.md:
```diff
@@ -22,26 +30,78 @@ browzine.script.src = "https://s3.amazonaws.com/browzine-adapters/summon/browzin
-  articleWording: "View Complete Issue",
-  journalBrowZineWebLinkText: "Browse Now",
-  articleBrowZineWebLinkText: "Browse Now",
+  summonJournalWording: "View the Journal",
```
Definitely cleaner naming for folks just dropping this into their systems I think 👍

karlbecker commented on package.json:
```diff
@@ -9,8 +9,14 @@
     "test": "karma start",
-    "deploy-staging": "AWS_BUCKET_FOLDER=summon-staging deploy-web-to-s3 src",
-    "deploy-production": "AWS_BUCKET_FOLDER=summon deploy-web-to-s3 src"
+    "deploy-summon-hotfix": "AWS_BUCKET_FOLDER=summon/hotfix deploy-web-to-s3 src/summon",
```
Thinking ahead! 👍

karlbecker commented:
Not sure how my review went through without a description at the top - but yeah, this looks good @tiandavis ! I'll reorganize that README, but in general everything is looking good.  Merge whenever you're ready - including if you want to wait for John to try it out some next week.

karlbecker commented:
@tiandavis - check out the reorganized README. 👍 ? Feel free to modify some more if you see additional improvements! 

Also, one thing I noticed: "View Journal in BrowZine" is probably not quite the default text we'll want to have for 360 Core, but I don't quite know what is best (maybe same as Summon?), so we can just tweak that when John comes back.

Thanks for all your hard work on this Tian!

tiandavis commented:
Added following feedback from @johnseguin:
- Added checks in Summon and 360 Core for browzineEnabled flag when determining whether to enhance Journal Cover.
- Added Journal Cover drop shadow treatment to 360 Core enhanced Journal Covers.

johnseguin commented:
@tiandavis  Is there a live demo link for Primo here?

tiandavis commented:
Hi @johnseguin! Here's the live Primo link:
https://sandbox01-na.alma.exlibrisgroup.com/discovery/search?query=any,contains,new%20england%20journal%20of%20medicine&tab=Everything&search_scope=MyInst_and_CI&vid=01COLSCHL_INST:BRZNTEST&offset=0

karlbecker commented:
Hey @johnseguin ,
Since Primo is being moved into this repo's codebase now, and setup will be a little different than with the old Primo, what should we say about [the old primo library here](https://github.com/thirdiron/primo-browzine)?

I think a big header like:

##  This code is deprecated; please check out the [new and improved version here](https://github.com/thirdiron/browzine-summon-adapter/pull/15).

at the top of primo-browzine would be fine, perhaps along with just removing most of the rest of the README text, possibly linking to an old commit of it if someone wants to know the old directions.

We can also change the name of this repo to something more broad than browzine-summon-adapter, since now it's a 3-in-1 tool! 🎉

johnseguin commented:
@karlbecker Yep - I think that sounds great.  Thanks!

tiandavis commented:
Hey Folks! Merging this to the `develop` branch. That should get us a release to respective staging folders on s3 and a stable location for testing partners.
